src/SAWScript/Crucible/LLVM/X86.hs

```diff
@@ -619,6 +658,117 @@ setupSimpleLoopFixpointFeature sym sc sawst cfg mvar func =
+               && not (List.isPrefixOf "cundefined" $ show $ W4.printSymExpr variable)
+               && not (List.isPrefixOf "calign_amount" $ show $ W4.printSymExpr variable)
+               && not (List.isPrefixOf "cnoSatisfyingWrite" $ show $ W4.printSymExpr variable)
+             )
+             implicit_params
+       body_tms <- mapM (viewSome $ toSC sym sawst) filtered_implicit_params
+       implicit_params' <- mapM (scExtCns sc) $ Set.toList $ foldMap getAllExtSet body_tms
+       initial_exprs <-
+         forM subst_pairs $
+           \ (MapF.Pair _var (SimpleInvariant.InvariantEntry initVal _current)) ->
+               toSC sym sawst initVal
+       current_exprs <-
+         forM subst_pairs $
+           \ (MapF.Pair _var (SimpleInvariant.InvariantEntry _init current)) ->
+               toSC sym sawst current
+
+       initial_tuple <- scTuple sc initial_exprs
+       current_tuple <- scTuple sc current_exprs
+
+       -- use the provided logging function to print the discovered
+       -- implicit parameters
+       when (phase == SimpleInvariant.InitialInvariant) $
+         do printFn "Loop invariant implicit parameters!"
+            forM_ implicit_params' $ \x ->
+                do printFn (show (ppTerm Verifier.SAW.SharedTerm.defaultPPOpts x))
+                   tp <- scTypeOf sc x
+                   printFn (show (ppTerm Verifier.SAW.SharedTerm.defaultPPOpts tp))
+
+       -- actually apply the arguments to the given term
+       inv <- scApplyAll sc (ttTerm func) (implicit_params' ++ [initial_tuple, current_tuple])
+
+       -- check that the produced term is type-correct
+       res <- scTypeCheck sc Nothing inv
+       case res of
+         Left _tcErr ->
+           do tpType <- scTypeOf sc initial_tuple
+              fail $ unlines [ "Loop invariant has incorrect type! State tuple has type:"
+                             , show (ppTerm Verifier.SAW.SharedTerm.defaultPPOpts tpType)
+                             ]
+         Right tp ->
+           do ok <- scConvertible sc True tp =<< scBoolType sc
+              unless ok $
+                fail $ unlines [ "Loop invaraiant must return a boolean value, but got:"
+                               , show (ppTerm Verifier.SAW.SharedTerm.defaultPPOpts tp)
+                                  -- TODO, get ppOpts from the right place
+                               ]
+       b <- bindSAWTerm sym sawst W4.BaseBoolRepr inv
+
+       -- Add goal metadata for the initial and inductive invariants
+       case phase of
+         SimpleInvariant.HypotheticalInvariant -> return b
+         SimpleInvariant.InitialInvariant ->
+           do (ann,b') <- W4.annotateTerm sym b
+              loc <- W4.getCurrentProgramLoc sym
+              let md = MS.ConditionMetadata
+                       { MS.conditionLoc = loc
+                       , MS.conditionTags = Set.singleton "initial loop invariant"
```

Comment:
It may be worth mentioning these tags in the help text, for the sake of discoverability.